Scrapper for course data

PetterRein opened the pull request:
- add scraper for course data from "https://www.ntnu.no/studier/emner/" 

- add also if the exam from was digital or not
 
- updates faculties

- fix minor quality of life improvements

PetterRein committed:
bump django version to 3.0


Wrong/Nonexisting kwargs used

PetterRein committed:
add search on grade and course adminpanel

PetterRein committed:
update names and add new faculty

Artorp committed:
Warn instead of exceptions, small changes to sleep, and add docs

PetterRein committed:
+1 on input

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+                # skip continuation exam
+                #is_ordinary = exam_status.find("abbr").text.strip() == "ORD"
+                #if not is_ordinary:
+                #    continue
```
Could we remove or handle this before merging?

PetterRein replied:
Yes

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+                term_is_digital_dict[term_std] = system.text.strip() == "INSPERA"
+            result[year] = term_is_digital_dict
+            # sleep a little bit to avoid hammering the website
+            time.sleep(0.05)
```
Shoudl we use the `sleep_time_mean_ms` here?

PetterRein replied:
Probably, was testing with time.sleep

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+        if (results[year].get(semester_code)):
+            return True
+        else:
+            return False
```
Could this be reduced to just this?
``` python
return semester_code in results[year]
```

PetterRein replied:
Jupp

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+      for key, value in exams_year.items(): 
+         if True == value: 
+             return True
+    return False
```
Inconsistent casing

PetterRein replied:
casing? line 118?

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+
+
+if __name__ == '__main__':
+    _test_functions()
```
Could we remove this before merging?

PetterRein replied:
Yup

oleast commented on grades.py:
```diff
@@ -44,11 +46,13 @@ def login(username, password):
-    base_url = "http://www.ime.ntnu.no/api/course/"
-    resp = requests.get(url=base_url + code)
+    #base_url = "http://www.ime.ntnu.no/api/course/"
+    #resp = requests.get(url=base_url + code)
```
Should probably remove these old API thingies?

oleast commented on grades.py:
```diff
@@ -44,11 +46,13 @@ def login(username, password):
     if not resp:
         return None
-    data = json.loads(resp.text)
+    #data = json.loads(resp.text)
```
Remove?

oleast commented on course_is_digital.py:
```diff
@@ -0,0 +1,142 @@
+        return result
+
+
+def course_have_digital_exam_semester(course_code: str, year: str, semester_code: str):
```
Should probably be `course_has_digital_exam`?

PetterRein replied:
Dont think so

oleast commented on grades/models.py:
```diff
@@ -40,16 +40,20 @@ def course_level(self):
+    def __str__(self):
+        return self.code
+
+    # FIX Permalink depercated
```
Do you know what was the actual use for this?

PetterRein replied:
Not really, only that it was used for reverse the URL of courses

oleast commented on scrapeCourse.py:
```diff
@@ -0,0 +1,140 @@
+    courseLevelText = facts_about_course[3].split("\n")[0][1:len(facts_about_course[3].split("\n")[0])]
+    study_level = courseLevel(courseLevelText, code)
+    last_year_taught = 0
+    taught_from = 2008
```
Why is this set to 2008?

PetterRein replied:
Every course on "emner" starts from 2008, but I know some courses started before that but we cannot get the data. taught_from is not used anywhere so maybe we could remove it

oleast commented on grades/models.py:
```diff
@@ -40,16 +40,20 @@ def course_level(self):
     semester_code = models.CharField("Semester", max_length=10)
     
     average_grade = models.FloatField()
+    digitalExam = models.BooleanField(default=False)
```
Should be `digital_exam` to keep up with Python naming practices.

PetterRein replied:
Yupp

oleast commented on scrapeCourse.py:
```diff
@@ -0,0 +1,140 @@
+from bs4 import BeautifulSoup
+import requests
+
+def getCourseData(code, faculty):
```
This file has inconsistent indentation with the rest of the project

PetterRein replied:
Will fix

oleast reviewed (commented):
Awesome work!! :heart_eyes: 

I added a few comments, but it is mostly nitpicking :smile: 

I have not been able to test it for myself yet, but I'll have a look at it when I can :smiley:

PetterRein committed:
Add functions for checking if course is digital exam

cleanup file

PetterRein committed:
add scrapper to get data about courses now that the api is down


[WIP] scrape course information


ff


dd


asdf


adf


fix wrong order of print and return


remove old api call


indentation

PetterRein committed:
add if exam was digital


delete .vscode files


fix django name convention


cleanup


cleanup


obs removed some functions in cleanup


more cleanup


add some comments


typo


update migration

PetterRein committed:
update function names

PetterRein committed:
add more data to course and grade

PetterRein committed:
try except check if exam was digital

oleast reviewed (approved):
Looks good to me :heart_eyes:

Artorp committed:
Cleanup some code

PetterRein committed:
add userAgent to fake servers

PetterRein committed:
add check for empty dict, even we know it was a exam then

PetterRein committed:
add some retry on gets if fail, with backof time if
